calendar.1
.\" Copyright (c) 1989, 1990, 1993
.\"     The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"     This product includes software developed by the University of
.\"     California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)calendar.1  8.1 (Berkeley) 6/29/93
.\" $FreeBSD$
.\"
.Dd June 13, 2002
.Dt CALENDAR 1
.Os
.Sh NAME
.Nm calendar
.Nd reminder service
.Sh SYNOPSIS
.Nm
.Op Fl a
.Op Fl A Ar num
.Op Fl B Ar num
.Op Fl F Ar friday
.Op Fl f Ar calendarfile
.Oo
.Bk -words
.Fl t Ar dd Ns
.Sm off
.Op . Ar mm Op . Ar year
.Sm on
.Ek
.Oc
.Op Fl W Ar num
.Op Fl U Ar UTC-offset
.Op Fl l Ar longitude
.Sh DESCRIPTION
The
.Nm
utility checks the current directory for a file named
.Pa calendar
and displays lines that begin with either today's date
or tomorrow's.
On the day before a weekend (normally Friday), events for the next
three days are displayed.
.Pp
The following options are available:
.Bl -tag -width Ds
.It Fl A Ar num
Print lines from today and the next
.Ar num
days (forward, future).
.It Fl a
Process the ``calendar'' files of all users and mail the results
to them.
This requires super-user privileges.
.It Fl B Ar num
Print lines from today and the previous
.Ar num
days (backward, past).
.It Fl F Ar friday
Specify which day of the week is ``Friday'' (the day before the
weekend begins).
Default is 5.
.It Fl f Pa calendarfile
Use
.Pa calendarfile
as the default calendar file.
.It Xo Fl t
.Sm off
.Ar dd
.Op . Ar mm Op . Ar year
.Sm on
.Xc
For test purposes only: set date directly to argument values.
.It Fl l Ar longitude , Fl U Ar UTC-offset
Only one is needed:
Perform lunar and solar calculations from this longitude or from
this UTC offset.
If neither is specified, the calculations will be based on the
difference between UTC time and localtime.
.It Fl W Ar num
Print lines from today and the next
.Ar num
days (forward, future).
Ignore weekends when calculating the number of days.
.El
.Pp
To handle calendars in your national code table you can specify
.Dq LANG=<locale_name>
in the calendar file as early as possible.
.Pp
To handle the local name of sequences, you can specify them as:
.Dq SEQUENCE=<first> <second> <third> <fourth> <fifth> <last>
in the calendar file as early as possible.
.Pp
The names of the following special days are recognized:
.Bl -tag -width 123456789012345 -compact
.It Easter
Catholic Easter.
.It Paskha
Orthodox Easter.
.It NewMoon
The lunar New Moon.
.It FullMoon
The lunar Full Moon.
.It MarEquinox
The solar equinox in March.
.It JunSolstice
The solar solstice in June.
.It SepEquinox
The solar equinox in March.
.It DecSolstice
The solar solstice in December.
.It ChineseNewYear
The first day of the Chinese year.
.El
These names may be reassigned to their local names via an assignment
like
.Dq Easter=Pasen
in the calendar file.
.Pp
Other lines should begin with a month and day.
They may be entered in almost any format, either numeric or as character
strings.
If the proper locale is set, national month and weekday
names can be used.
A single asterisk (``*'') matches every month.
A day without a month matches that day of every week.
A month without a day matches the first of that month.
Two numbers default to the month followed by the day.
Lines with leading tabs default to the last entered date, allowing
multiple line specifications for a single date.
.Pp
The names of the recognized special days may be followed by a
positive or negative integer, like:
.Dq Easter+3
or
.Dq Pashka-4 .
.Pp
Weekdays may be followed by ``-4'' ...\& ``+5'' (aliases for
last, first, second, third, fourth) for moving events like
``the last Monday in April''.
.Pp
By convention, dates followed by an asterisk are not fixed, i.e., change
from year to year.
.Pp
Day descriptions start after the first <tab> character in the line;
if the line does not contain a <tab> character, it is not displayed.
If the first character in the line is a <tab> character, it is treated as
a continuation of the previous line.
.Pp
The ``calendar'' file is preprocessed by
.Xr cpp 1 ,
allowing the inclusion of shared files such as lists of company holidays or
meetings.
If the shared file is not referenced by a full pathname,
.Xr cpp 1
searches in the current (or home) directory first, and then in the
directory
.Pa /usr/share/calendar .
Empty lines and lines protected by the C commenting syntax
.Pq Li /* ... */
are ignored.
.Pp
Some possible calendar entries (<tab> characters highlighted by
\fB\et\fR sequence)
.Bd -unfilled -offset indent
LANG=C
Easter=Ostern

#include <calendar.usholiday>
#include <calendar.birthday>

6/15\fB\et\fRJune 15 (if ambiguous, will default to month/day).
Jun. 15\fB\et\fRJune 15.
15 June\fB\et\fRJune 15.
Thursday\fB\et\fREvery Thursday.
June\fB\et\fREvery June 1st.
15 *\fB\et\fR15th of every month.

May Sun+2\fB\et\fRsecond Sunday in May (Muttertag)
04/SunLast\fB\et\fRlast Sunday in April,
\fB\et\fRsummer time in Europe
Easter\fB\et\fREaster
Ostern-2\fB\et\fRGood Friday (2 days before Easter)
Paskha\fB\et\fROrthodox Easter

.Ed
.Sh FILES
.Bl -tag -width calendar.christian -compact
.It Pa calendar
file in current directory
.It Pa ~/.calendar
.Pa calendar
HOME directory.
A chdir is done into this directory if it exists.
.It Pa ~/.calendar/calendar
calendar file to use if no calendar file exists in the current directory.
.It Pa ~/.calendar/nomail
do not send mail if this file exists.
.El
.Pp
The following default calendar files are provided in
.Pa /usr/share/calendars:
.Pp
.Bl -tag -width calendar.southafrica -compact
.It Pa calendar.all
File which includes all the default files.
.It Pa calendar.australia
Calendar of events in Australia.
.It Pa calendar.birthday
Births and deaths of famous (and not-so-famous) people.
.It Pa calendar.christian
Christian holidays.
This calendar should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year.
.It Pa calendar.computer
Days of special significance to computer people.
.It Pa calendar.croatian
Calendar of events in Croatia.
.It Pa calendar.dutch
Calendar of events in the Netherlands.
.It Pa calendar.freebsd
Birthdays of
.Fx
committers.
.It Pa calendar.french
Calendar of events in France.
.It Pa calendar.german
Calendar of events in Germany.
.It Pa calendar.history
Everything else, mostly U.S.\& historical events.
.It Pa calendar.holiday
Other holidays, including the not-well-known, obscure, and
.Em really
obscure.
.It Pa calendar.judaic
Jewish holidays.
This calendar should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year.
.It Pa calendar.music
Musical events, births, and deaths.
Strongly oriented toward rock 'n' roll.
.It Pa calendar.newzealand
Calendar of events in New Zealand.
.It Pa calendar.russian
Russian calendar.
.It Pa calendar.southafrica
Calendar of events in South Africa.
.It Pa calendar.usholiday
U.S.\& holidays.
This calendar should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year.
.It Pa calendar.world
Includes all calendar files except for national files.
.El
.Sh COMPATIBILITY
The
.Nm
program previously selected lines which had the correct date anywhere
in the line.
This is no longer true, the date is only recognized when it occurs
at the beginning of a line.
.Sh SEE ALSO
.Xr at 1 ,
.Xr cpp 1 ,
.Xr mail 1 ,
.Xr cron 8
.Sh HISTORY
A
.Nm
command appeared in
.At v7 .
.Sh NOTES
Chinese New Year is calculated at 120 degrees east of Greenwich,
which roughly corresponds with the east coast of China.
For people west of China, this might result that the start of Chinese
New Year and the day of the related new moon might differ.
.Pp
The phases of the moon and the longitude of the sun are calculated
against the local position which corresponds with 30 degrees times
the time-difference towards Greenwich.
.Pp
The new and full moons are happening on the day indicated: They
might happen in the time period in the early night or in the late
evening.
It doesn't indicate that they are starting in the night on that date.
.Sh BUGS
The
.Nm
utility does not handle Jewish holidays.
.Pp
There is no possibility to properly specify the local position
needed for solar and lunar calculations.


calendar.c
/*-
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1989, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif

#if 0
#ifndef lint
static char sccsid[] = "@(#)calendar.c  8.3 (Berkeley) 3/25/94";
#endif
#endif

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include <err.h>
#include <errno.h>
#include <locale.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "calendar.h"

#define	UTCOFFSET_NOTSET	100	/* Expected between -24 and +24 */
#define	LONGITUDE_NOTSET	1000	/* Expected between -360 and +360 */

struct passwd	*pw;
int		doall = 0;
int		debug = 0;
char		*DEBUG = NULL;
double		UTCoffset = UTCOFFSET_NOTSET;
int		eastlongitude = LONGITUDE_NOTSET;
time_t		f_time = 0;

int	f_dayAfter = 0;		/* days after current date */
int	f_dayBefore = 0;	/* days before current date */
int	Friday = 5;		/* day before weekend */

static void	usage(void) __dead2;

int
main(int argc, char *argv[])
{
	int ch;

	(void)setlocale(LC_ALL, "");

	while ((ch = getopt(argc, argv, "-A:aB:dD:F:f:l:t:U:W:")) != -1)
		switch (ch) {
		case '-':		/* backward contemptible */
		case 'a':
			if (getuid()) {
				errno = EPERM;
				err(1, NULL);
			}
			doall = 1;
			break;

		case 'f': /* other calendar file */
			calendarFile = optarg;
			break;

		case 'W': /* we don't need no steenking Fridays */
			Friday = -1;
			/* FALLTHROUGH */

		case 'A': /* days after current date */
			f_dayAfter = atoi(optarg);
			break;

		case 'B': /* days before current date */
			f_dayBefore = atoi(optarg);
			break;

		case 'F': /* Change the time: When does weekend start? */
			Friday = atoi(optarg);
			break;
		case 'l': /* Change longitudal position */
			eastlongitude = strtol(optarg, NULL, 10);
			break;
		case 'U': /* Change UTC offset */
			UTCoffset = strtod(optarg, NULL);
			break;

		case 'd':
			debug = 1;
			break;
		case 'D':
			DEBUG = optarg;
			break;
		case 't': /* other date, undocumented, for tests */
			f_time = Mktime(optarg);
			break;

		case '?':
		default:
			usage();
		}

	argc -= optind;
	argv += optind;

	if (argc)
		usage();

	/* use current time */
	if (f_time <= 0)
		(void)time(&f_time);

	/* if not set, determine where I could be */
	{
		if (UTCoffset == UTCOFFSET_NOTSET &&
		    eastlongitude == LONGITUDE_NOTSET) {
			/* Calculate on difference between here and UTC */
			time_t t;
			struct tm tm;
			long utcoffset, hh, mm, ss;
			double uo;

			time(&t);
			localtime_r(&t, &tm);
			utcoffset = tm.tm_gmtoff;
			/* seconds -> hh:mm:ss */
			hh = utcoffset / SECSPERHOUR;
			utcoffset %= SECSPERHOUR;
			mm = utcoffset / SECSPERMINUTE;
			utcoffset %= SECSPERMINUTE;
			ss = utcoffset;

			/* hh:mm:ss -> hh.mmss */
			uo = mm + (100.0 * (ss / 60.0));
			uo /=  60.0 / 100.0;
			uo = hh + uo / 100;

			UTCoffset = uo;
			eastlongitude = UTCoffset * 15;
		} else if (UTCoffset == UTCOFFSET_NOTSET) {
			/* Base on information given */
			UTCoffset = eastlongitude / 15;
		} else if (eastlongitude == LONGITUDE_NOTSET) {
			/* Base on information given */
			eastlongitude = UTCoffset * 15;
		}
	}

	settimes(f_time, f_dayBefore, f_dayAfter, &tp1, &tp2);
	generatedates(&tp1, &tp2);

	/*
	 * FROM now on, we are working in UTC.
	 * This will only affect moon and sun related events anyway.
	 */
	if (setenv("TZ", "UTC", 1) != 0)
		errx(1, "setenv: %s", strerror(errno));
	tzset();

	if (debug)
		dumpdates();

	if (DEBUG != NULL) {
		dodebug(DEBUG);
		exit(0);
	}

	if (doall)
		while ((pw = getpwent()) != NULL) {
			(void)setegid(pw->pw_gid);
			(void)initgroups(pw->pw_name, pw->pw_gid);
			(void)seteuid(pw->pw_uid);
			if (!chdir(pw->pw_dir))
				cal();
			(void)seteuid(0);
		}
	else
		cal();
	exit(0);
}


static void __dead2
usage(void)
{

	fprintf(stderr, "%s\n%s\n%s\n",
	    "usage: calendar [-a] [-A days] [-B days] [-F friday] "
	    "[-f calendarfile]",
	    "                [-d] [-t dd[.mm[.year]]] [-W days]",
	    "                [-U utcoffset] [-l longitude]"
	    );
	exit(1);
}
